Tools/cases_generator/test_generator.py

```diff
@@ -0,0 +1,310 @@
+    output = """
+        TARGET(OP) {
+            PyObject *value = PEEK(1);
+            spam();
+            STACK_SHRINK(1);
+            DISPATCH();
+        }
+    """
+    run_test(input, output)
+
+def test_inst_one_push():
+    input = """
+        inst(OP, (-- res)) {
+            spam();
+        }
+    """
+    output = """
+        TARGET(OP) {
+            PyObject *res;
+            spam();
+            STACK_GROW(1);
+            POKE(1, res);
+            DISPATCH();
+        }
+    """
+    run_test(input, output)
+
+def test_inst_one_push_one_pop():
+    input = """
+        inst(OP, (value -- res)) {
+            spam();
+        }
+    """
+    output = """
+        TARGET(OP) {
+            PyObject *value = PEEK(1);
+            PyObject *res;
+            spam();
+            POKE(1, res);
+            DISPATCH();
+        }
+    """
+    run_test(input, output)
+
+def test_binary_op():
+    input = """
+        inst(OP, (left, right -- res)) {
+            spam();
+        }
+    """
+    output = """
+        TARGET(OP) {
+            PyObject *right = PEEK(1);
+            PyObject *left = PEEK(2);
+            PyObject *res;
+            spam();
+            STACK_SHRINK(1);
+            POKE(1, res);
+            DISPATCH();
+        }
+    """
+    run_test(input, output)
+
+def test_predictions():
+    input = """
+        inst(OP1, (--)) {
+        }
+        inst(OP2, (--)) {
+        }
+        inst(OP3, (--)) {
+            DEOPT_IF(xxx, OP1);
+            PREDICT(OP2);
+        }
+    """
+    output = """
+        TARGET(OP1) {
+            PREDICTED(OP1);
+            DISPATCH();
+        }
+
+        TARGET(OP2) {
+            PREDICTED(OP2);
+            DISPATCH();
+        }
+
+        TARGET(OP3) {
+            DEOPT_IF(xxx, OP1);
+            PREDICT(OP2);
+            DISPATCH();
+        }
+    """
+    run_test(input, output)
+
+def test_error_if_plain():
+    input = """
+        inst(OP, (--)) {
+            ERROR_IF(cond, label);
+        }
+    """
+    output = """
+        TARGET(OP) {
+            if (cond) goto label;
+            DISPATCH();
+        }
+    """
+    run_test(input, output)
+
+def test_error_if_pop():
+    input = """
+        inst(OP, (left, right -- res)) {
+            ERROR_IF(cond, label);
+        }
+    """
+    output = """
+        TARGET(OP) {
+            PyObject *right = PEEK(1);
+            PyObject *left = PEEK(2);
+            PyObject *res;
+            if (cond) goto pop_2_label;
+            STACK_SHRINK(1);
+            POKE(1, res);
+            DISPATCH();
+        }
+    """
+    run_test(input, output)
+
+def test_cache_effect():
+    input = """
+        inst(OP, (counter/1, extra/2, value --)) {
+        }
+    """
+    output = """
+        TARGET(OP) {
+            PyObject *value = PEEK(1);
+            uint16_t counter = read_u16(&next_instr[0].cache);
+            uint32_t extra = read_u32(&next_instr[1].cache);
+            STACK_SHRINK(1);
```

Comment:
I wonder if there's a micro-optimisation here to put the PEEK(1) just before the STACK_SHINK (more locality).

Reply:
I don't think so. The PEEK(1) is part of the prologue while the STACK_SHRINK(1) is part of the epilogue. The whole block of code goes in between (though here that's empty). See other examples e.g. test_inst_one_pop() at line 58.